[FLINK-18515][Kinesis] Adding FanOutRecordPublisher for Kinesis EFO support

dannycranmer opened the pull request:
## What is the purpose of the change

Adding `FanOutRecordPublisher`, a new implementation of `RecordPublisher` to add support for Enhanced Fan Out (EFO) from Kinesis Streams using the `FlinkKinesisConsumer`.

## Brief change log

- `FanOutRecordPublisher`
  - Consumes records from Kinesis using EFO via `KinesisV2Proxy`
  - `FanOutShardSubscriber` used to split consumption and consumer processing and error handling
- `ShardConsumer`
  - Starting position from `LATEST` is converted to `AT_TIMESTAMP` `now()` to prevent data loss if error occurs on first read
- `RecordPublisherFactory` 
  - Added a `close()` method to allow `KinesisAsynClient` to be closed on shutdown
- `FullJitterBackoff`
  - Pulled the jitter backoff implementation into a separate class and referenced throughout
-  `KinesisProxyV2Interface` / `KinesisProxyV2`
  - Added `SubscribeToShard` thin proxy layer
- `AwsV2Util`
  - Updated client configurations for EFO:
    - `maxConcurrency`: Added a new configuration to allow EFO concurrency to be defined separately from other services (list shards etc)
    - `protocol`: Setting to `HTTP2`
    - `connectionAcquisitionTimeout`/`healthCheckPingPeriod`/`initialWindowSize` set to handle startup with large shard count

## Verifying this change

This change added tests and can be verified as follows:
- `FlinkKinesisConsumerTest`
- `KinesisDataFetcherTest`
- `ShardConsumerFanOutTest`
- `ShardConsumerTest`
- `FanOutRecordPublisherConfigurationTest`
- `FanOutRecordPublisherTest`
- `KinesisProxyV2Test`
- `AwsV2UtilTest`
- `KinesisConfigUtilTest`

Fake behaviours for EFO added:
- `FakeKinesisFanOutBehavioursFactory`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes (added EFO consumption)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? There is a [follow up task](https://issues.apache.org/jira/browse/FLINK-18870) for documentation

tzulitai commented:
I'll start the review of this after we've resolved the comments in #12881.

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java:
```diff
@@ -330,23 +359,26 @@ public KinesisDataFetcher(List<String> streams,
+			final List<KinesisStreamShardState> subscribedShardsState,
+			final HashMap<String, String> subscribedStreamsToLastDiscoveredShardIds,
+			final FlinkKinesisProxyFactory kinesisProxyFactory,
+			@Nullable final FlinkKinesisProxyV2Factory kinesisProxyV2Factory) {
```
👍

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java:
```diff
@@ -389,25 +424,54 @@ private RecordEmitter createRecordEmitter(Properties configProps) {
+		switch (recordPublisherType) {
+			case EFO:
+				return new FanOutRecordPublisherFactory(kinesisProxyV2Factory.create(configProps));
+			case POLLING:
```
@xiaolong-sn How so? Please elaborate

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutRecordPublisher.java:
```diff
@@ -0,0 +1,245 @@
+			nextStartingPosition.getStartingMarker());
+
+		Consumer<SubscribeToShardEvent> eventConsumer = event -> {
+			RecordBatch recordBatch = new RecordBatch(toSdkV1Records(event.records()), subscribedShard, event.millisBehindLatest());
```
@xiaolong-sn I have verified that the `data` is passed by reference, it is not copied. So the overhead would be the GC on the new Object inflation. The rest is shuffling references. Once we update Polling to SDK V2 we can use KCL v2.x and get rid of the transform. The alternative now is to import KCL v2.x which would bloat the dependency chain and application size.

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutRecordPublisherFactory.java:
```diff
@@ -0,0 +1,95 @@
+		String stream = streamShardHandle.getStreamName();
+		FanOutRecordPublisherConfiguration configuration = new FanOutRecordPublisherConfiguration(consumerConfig, singletonList(stream));
+
+		Optional<String> streamConsumerArn = configuration.getStreamConsumerArn(stream);
```
@xiaolong-sn You can pass in stream consumer via properties, effectively using the NONE configuration option

dannycranmer commented on flink-connectors/flink-connector-kinesis/pom.xml:
```diff
@@ -161,6 +163,30 @@ under the License.
+			<version>${aws.sdkv2.version}</version>
+		</dependency>
+
+		<dependency>
```
@xiaolong-sn I took a look into the error you reported. You were getting a Jackson error:

```
Caused by: java.lang.NoSuchFieldError: NO_INTS
	at com.fasterxml.jackson.dataformat.cbor.CBORParser.<init>(CBORParser.java:285)
```

Which is similar to [this error on stackoverflow](https://stackoverflow.com/questions/60756770/error-java-lang-nosuchfielderror-no-ints)

And you fixed it by using:

```
<dependency>
    <groupId>com.fasterxml.jackson.dataformat</groupId>
    <artifactId>jackson-dataformat-cbor</artifactId>
    <version>2.6.7</version>
</dependency>
```

I cannot replicate this issue, I can see that `jackson-dataformat-cbor` is pulled through three chains in a bare applications:
- Flink @ 2.10.1 (https://github.com/apache/flink/blob/master/pom.xml#L126)
- AWS SDK v1 @ 2.6.7 (https://github.com/aws/aws-sdk-java/blob/master/pom.xml#L275)
- AWS SDK v2 @ 2.10.4 (https://github.com/aws/aws-sdk-java-v2/blob/master/pom.xml#L87)

It is using the Flink managed version currently:
- `com.fasterxml.jackson.dataformat:jackson-dataformat-cbor:jar:2.10.1:compile`

I tried explicitly adding `2.6.7` but it seems this is banned:

```
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.BannedDependencies failed with message:
Found Banned Dependency: com.fasterxml.jackson.dataformat:jackson-dataformat-cbor:jar:2.6.7
```

Can you please provide a minimal application where I can reproduce and test the problem?

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyV2.java:
```diff
@@ -30,15 +34,28 @@
+
+	@Override
+	public void close() {
+		kinesisAsyncClient.close();
```
@tzulitai I later realised that I need to close the Async HTTP Client used by this client explicitly. This has been done in the next PR (registration). If you want me to pull that back into this PR let me know. The result of not closing the HTTP Client is that the thread pool is not shutdown when the job is cancelled/errors.

tzulitai commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutShardSubscriber.java:
```diff
@@ -0,0 +1,466 @@
+				// Empty the queue and add a poison pill to terminate this subscriber
+				// The synchronized block ensures that new data is not written in the meantime
+				queue.clear();
+				enqueueEvent(new SubscriptionErrorEvent(throwable));
```
In the case of backpressure in Flink's network stack, would this also mean that job failure is postponed until this `SubscriptionErrorEvent` is postponed until the source connector thread consumes it?

dannycranmer replied:
Yes. Most errors are recoverable. But if a non-recoverable error is thrown and application is currently consuming a record batch. The application would finish processing the record batch, then receive the error and cancel the Flink job. I think this is ok, do you have concerns?

In the case of multiple shards, typically a non-recoverable error from the network stack would be detected by all `ShardConsumers` and it would be a race as to who cancels first. I am thinking of typical errors such as:
- Stream gets deleted
- Stream Consumer get deleted
- Credentials are invalid

tzulitai reviewed (commented):
Thanks for the great work @dannycranmer!
I've finished a rather thorough first pass, and am mostly trying to understand some scenarios around job cancellation with the current inter-thread communication model implemented in the `FanOutShardSubscriber`.

Overall the code is good, just want to clarify some of my comments before we proceed with merging this.

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java:
```diff
@@ -102,6 +107,11 @@ public void run() {
 		try {
 			while (isRunning()) {
 				final RecordPublisherRunResult result = recordPublisher.run(batch -> {
+					if (!batch.getDeaggregatedRecords().isEmpty()) {
```
I found this log useful while developing and also for diagnosing problems for running applications. It is `DEBUG` so would not clutter the logs normally. I am happy to remove this if you would prefer.

tzulitai replied:
No need to remove, lets leave that as is!

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutRecordPublisher.java:
```diff
@@ -0,0 +1,245 @@
+			configuration.getSubscribeToShardBaseBackoffMillis(),
+			configuration.getSubscribeToShardMaxBackoffMillis(),
+			configuration.getSubscribeToShardExpConstant(),
+			++attempt);
```
Yep, will do

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutShardSubscriber.java:
```diff
@@ -0,0 +1,466 @@
+		 */
+		private boolean enqueueEvent(final FanOutSubscriptionEvent event) {
+			try {
+				if (!queue.offer(event, ENQUEUE_WAIT_SECONDS, SECONDS)) {
```
Yes correct. This is a measure to prevent lock up in high back-pressure scenarios. If the app is applying large back-pressure and `onNext` is blocked on queue push, this 5 second retry allows `onError` to set `cancel = true` and then slip the error event into the queue. The downside to this is that we might end up throwing away 2 record batches and re-requesting them on next subscription. However my testing revealed that the AWS SDK Async client and Netty client do not like these threads hanging around and it is happier this way.

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutShardSubscriber.java:
```diff
@@ -0,0 +1,466 @@
+ * </p>
+ */
+@Internal
+public class FanOutShardSubscriber {
```
Where do you see these warning? I do not see them in terminal output, are they IntelliJ inspections?

dannycranmer replied:
Have addressed your comment but cannot see the warnings. If you share an inspections profile or point me at one (or docs) then that would be great. Thanks

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutShardSubscriber.java:
```diff
@@ -0,0 +1,466 @@
+
+		private Subscription subscription;
+
+		private boolean cancelled = false;
```
Yep, done.

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/AWSUtil.java:
```diff
@@ -277,7 +278,9 @@ public static void setAwsClientConfigProperties(ClientConfiguration config,
 	public static StartingPosition getStartingPosition(final SequenceNumber sequenceNumber, final Properties configProps) {
-		if (SENTINEL_AT_TIMESTAMP_SEQUENCE_NUM.get().equals(sequenceNumber)) {
+		if (sequenceNumber.equals(SENTINEL_LATEST_SEQUENCE_NUM.get())) {
+			return StartingPosition.fromTimestamp(new Date());
```
Done. Also I removed some redundant code as a result of this in `PollingRecordPublisher`. This means that an additional call to `GetShardIterator` will be performed when that condition was previously `true`.

dannycranmer commented on flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutShardSubscriber.java:
```diff
@@ -0,0 +1,466 @@
+ *      <li>1 event being added to the queue by the network (blocking)</li>
+ *   </ul>
+ * </p>
+ */
```
:+1:

dannycranmer commented:
@tzulitai I have addressed all comments. As you will be merging them with the registration pull request I have pushed to that branch, to avoid having to rebase. If you would prefer I pushed here and rebased the other, please let me know.

tzulitai commented:
Had a final pass at the code, LGTM.
Thanks for addressing the comments @dannycranmer! Am merging this now ...
